Fix a bug in wxWMP10MediaBackend introduced by commit 294436c and add a comment that should help preventing such bugs in future

PBfordev opened the pull request:
I was aware of dangers of calling `wxBasicString` ctor/methods taking a `BSTR` with a string literal, yet in #502 I still forgot to take care about it in few places.

I suppose the potential dangers of `BSTR`/`wxString`/string literal ambiguity were the reasons why wxBasicString did not originally have a ctor/methods taking a `BSTR`. These bugs can happen with raw `BSTR`s too but with wxWidgets it is probably even trickier when (in Unicode UTF-16 build) `wxBasicString("abc")` is OK (results in calling `wxBasicString(const wxString&)` while `wxBasicString(wx{S|T}("abc"))` will result in a crash.

I have added some strongly worded comments to `wxBasicString` to assist in avoiding bugs like the one above. While the API documentation has never been enough to prevent someone from shooting himself in the foot, considering `wxBasicString` is a low-profile internal class, the chances of people misusing it are hopefully rather low.

The only other option would probably be, in the name of safety over convenience, to remove the ctor and method(s) taking a `BSTR` and have only `CreateFromBSTR()` (creates a copy of a `BSTR`) and `AttachBSTR()` (takes ownership of a `BSTR`). There would still be nothing to prevent anyone from calling call such a method with a wide string literal but the name could make it more obvious that a `BSTR` is expected.

PBfordev committed:
Add comments on how to avoid some pitfalls when using wxBasicString.

PBfordev committed:
Fix a wxBasicString-related bug in wxWMP10MediaBackend introduced by commit 294436c8.

vadz commented:
Yes, this is really dangerous... I think it would be better to have `{Create,Assign}From{String,BSTR}()` instead, especially as `CreateFromString("foo")` is not really worse than `wxBasicString(wxString("foo"))`.

pbfordev commented:
I have made some modifications to `wxBasicString` API. As I was not sure if they are exactly what you requested and did not want to pollute the PR with many pointless commits, I created a GIST:  https://gist.github.com/pbfordev/44a3f785e1a9b79139e0bc6635d19aaa

In particular, I was not sure about naming the methods and whether to have only the default and copy constructor or also keep a ctor taking a `wxString`. This rather convenient ctor was already present in the original `wxBasicString` and I consider it rather safe: The worst that could happen when misused is someone passing a `BSTR` to it, for some odd reason mistakenly thinking that `wxBasicString` will take its ownership.

Please, if you feel it will take you less time and effort to make any further changes by yourself, do not hesitate to do so, my feelings will not be hurt in any way. ;) As I wrote before, I wasted too much of your time on this simple and unimportant issue already...

vadz commented:
I was thinking of having static, factory-like `Create()` which would return the new `wxBasicString` from `BSTR`. If we keep ctor from `wxString`, we probably don't need `CreateFromString()` variant, so it would be enough to have `static wxBasicString CreateFromBSTR(BSTR)` method. Of course, it would be still possible to misuse it by doing `CreateFromBSTR("foo")`, but passing a literal string to a function with `BSTR` in its name seems much less likely to happen accidentally.

BTW, there is no problem at all with making as many commits as you want, we can (and probably will) always squash them later when merging.

PBfordev committed:
Make wxBasicString lean again

Keep wxBasicString API minimalist, providing only the methods actually used in wxWidgets code (and Get/Copy() for backwards compatibility).

pbfordev commented:
After having thought some more about `wxBasicString` and I now believe that I chose a wrong approach before, when I wanted to make `wxBasicString` a flexible fully-featured RAII wrapper for `BSTR`. 

`wxBasicString` is an internal, undocumented, and rarely used class and therefore its API should be very lean, providing only the methods that are used. The less code, the less bloat and lesser potential for bugs. If a need for a new method arises in the future, it would be easy to extend the API.

The latest commit reflects this approach.

PBfordev committed:
Remove invalid "wxBasicString::" qualification

Remove invalid "wxBasicString::" qualification from inline wxBasicString(wxString&) constructor.

vadz commented on include/wx/msw/ole/oleutils.h:
```diff
@@ -65,24 +65,26 @@ inline void wxOleUninitialize()
-    // Returns the owned BSTR and gives up its ownership
-    BSTR Detach();
+    // Creates the owned BSTR from a wxString
+    void AssignFromwxString(const wxString& str);
```
Could we call it just `AssignFromString()`? This reads better and it does take any string due to implicit conversions to `wxString`, after all.

vadz reviewed (approved):
I'm totally fine with this minimalistic approach, thanks.

PBfordev committed:
Rename wxBasicString's AssignFromwxString() to AssignFromString

The method can take any string due to implicit conversions to wxString.

PBfordev committed:
Fix a typo in wxBasicString comment

pbfordev commented:
I have changed the method name as requested.

Once again, sorry for wasting so much of your time on such a simple issue, thanks.

vadz commented:
Once again, no problem with taking too much time and, on the contrary, thanks for persevering and making this even better than initially planned!

Applied now, after squashing and then splitting into a couple of commits, please let me know if I did anything wrong and thanks again!
